File: ScaleformUI_Csharp/Menus/Pause Menus/Elements/Columns/PM_Column.cs
using ScaleformUI.Elements;
using ScaleformUI.PauseMenu;

namespace ScaleformUI.Menus
{
    public enum PM_COLUMNS
    {
        LEFT,
        MIDDLE,
        RIGHT,
        EXTRA3,
        EXTRA4,
        LEFT_MIDDLE,
        MIDDLE_RIGHT,
    };

    public delegate void IndexChanged(int index);
    public delegate void ItemSelect(TabView menu, TabLeftItem item, int leftItemIndex);
    public delegate void RightItemSelect(TabView menu, SettingsItem item, int leftItemIndex, int rightItemIndex);

    public class PM_Column
    {
        internal PM_COLUMNS position;
        internal int index = 0;
        //used in playerlisttab
        internal int type = -1;
        private bool columnVisible;

        internal bool visible => Parent != null && Parent.Visible && Parent.Parent != null && Parent.Parent.Visible;
        public List<PauseMenuItem> Items { get; internal set; } = new List<PauseMenuItem>();
        public virtual int Index
        {
            get => index;
            set
            {
                Items[index].Selected = false;
                index = value;
                if (index < 0)
                    index = Items.Count - 1;
                else if (index >= Items.Count)
                    index = 0;
                Items[index].Selected = true;
                if (visible && Parent.CurrentColumnIndex == (int)position)
                    Main.PauseMenu._pause.CallFunction("SET_COLUMN_HIGHLIGHT", (int)position, index, false, false);
            }
        }

        public bool ColumnVisible
        {
            get => columnVisible;
            set
            {
                columnVisible = value;
                Main.PauseMenu._pause.CallFunction("SHOW_COLUMN", (int)position, columnVisible);
            }
        }
        public int VisibleItems { get; internal set; }
        public bool Focused { get; internal set; }
        public string CaptionLeft { get; internal set; } = string.Empty;
        public string CaptionRight { get; internal set; } = string.Empty;
        public string Label { get; internal set; }
        public SColor Color { get; set; } = SColor.HUD_Freemode;
        public BaseTab Parent { get; internal set; }

        public PM_Column(PM_COLUMNS position)
        {
            this.position = position;
        }

        public PM_Column(int position) : this((PM_COLUMNS)position) { }

        public virtual void AddItem(PauseMenuItem item)
        {
            Items.Add(item);
        }

        public void Clear()
        {
            ClearColumn();
        }

        public virtual void Populate() { }

        public virtual void SetDataSlot(int index) { }

        public virtual void UpdateSlot(int index) { }

        public virtual void AddSlot(int index) { }
        public virtual void RemoveSlot(int idx)
        {
            if (idx >= Items.Count) return;
            var selectedItem = Index;
            Items[idx].Selected = false;
            Items.RemoveAt(idx);
            if (visible)
                Main.PauseMenu._pause.CallFunction("REMOVE_SLOT", (int)position, idx, false, false);
            if (Items.Count > 0)
            {
                if (idx == this.index)
                    index = idx >= Items.Count ? Items.Count - 1 : idx >= 0 && idx < Items.Count ? idx : 0;
                else
                {
                    if (selectedItem < Items.Count)
                        index = selectedItem;
                    else
                        index = Items.Count - 1;
                }
                Items[index].Selected = true;
                if (visible && Parent.CurrentColumnIndex == (int)position)
                    Main.PauseMenu._pause.CallFunction("SET_COLUMN_HIGHLIGHT", (int)position, idx, false, false);
            }
        }

        public virtual void GoUp() { }
        public virtual void GoDown() { }
        public virtual void GoLeft() { }
        public virtual void GoRight() { }
        public virtual void Select() { }
        public virtual void GoBack() { }
        public virtual void MouseScroll(int dir) { }

        public virtual void HighlightColumn(bool highlighted = false, bool moveFocus = false, bool prevHighlight = false)
        {
            if (visible)
                Main.PauseMenu._pause.CallFunction("SET_COLUMN_FOCUS", (int)position, highlighted, moveFocus, prevHighlight);
        }
        public virtual void ClearColumn()
        {
            Items.Clear();
            index = 0;
            if (visible)
                Main.PauseMenu._pause.CallFunction("SET_DATA_SLOT_EMPTY", (int)position);
        }
        public virtual void ShowColumn(bool show = true)
        {
             Main.PauseMenu._pause.CallFunction("DISPLAY_DATA_SLOT", (int)position);
        }
        public virtual void InitColumnScroll(bool visible, int columns, ScrollType scrollType, ScrollArrowsPosition arrowPosition, bool @override = false, float xColOffset = 0f)
        {
             Main.PauseMenu._pause.CallFunction("INIT_COLUMN_SCROLL", (int)position, visible, columns, (int)scrollType, (int)arrowPosition, @override, xColOffset);
        }
        public virtual void SetColumnScroll(int currentPosition, int maxPosition, int maxVisible, string caption, bool forceInvisible = false, string captionR = "")
        {
             Main.PauseMenu._pause.CallFunction("SET_COLUMN_SCROLL", (int)position, currentPosition, maxPosition, maxVisible, caption, forceInvisible, captionR);
        }

        public virtual void SetColumnScroll(int currentPosition, int maxPosition, int maxVisible = -1)
        {
            if (visible)
                Main.PauseMenu._pause.CallFunction("SET_COLUMN_SCROLL", (int)position, currentPosition, maxPosition, maxVisible);
        }

        public virtual void SetColumnScroll(string caption, params object[] args)
        {
            if (visible)
            {
                BeginScaleformMovieMethod(Main.PauseMenu._pause.Handle, "SET_COLUMN_SCROLL");
                ScaleformMovieMethodAddParamInt((int)position);
                ScaleformMovieMethodAddParamInt(0);
                ScaleformMovieMethodAddParamInt(0);
                ScaleformMovieMethodAddParamInt(0);
                BeginTextCommandScaleformString(caption);
                foreach (object arg in args)
                {
                    switch (arg)
                    {
                        case int:
                            AddTextComponentInteger((int)arg);
                            break;
                        case string:
                            AddTextComponentSubstringPlayerName((string)arg);
                            break;
                        case float:
                            AddTextComponentFloat((float)arg, 2);
                            break;
                    }
                }
                EndTextCommandScaleformString_2();
                EndScaleformMovieMethod();
            }
        }
        public virtual void SetColumnScroll(string caption, string rightC)
        {
            if (visible)
            {
                BeginScaleformMovieMethod(Main.PauseMenu._pause.Handle, "SET_COLUMN_SCROLL");
                ScaleformMovieMethodAddParamInt((int)position);
                ScaleformMovieMethodAddParamInt(0);
                ScaleformMovieMethodAddParamInt(0);
                ScaleformMovieMethodAddParamInt(0);
                BeginTextCommandScaleformString("CELL_EMAIL_BCON");
                AddTextComponentSubstringPlayerName(caption);
                EndTextCommandScaleformString_2();
                ScaleformMovieMethodAddParamBool(false);
                ScaleformMovieMethodAddParamPlayerNameString(rightC);
                EndScaleformMovieMethod();
            }
        }
    }
}

File: ScaleformUI_Lua/src/Menus/PauseMenus/Elements/Columns/_PM_Column.lua
PM_Column = {}

function PM_Column.New(pos)
    local data = {
        position = pos or -1,
        index = 1,
        type = -1, -- used in PlayerListTab :3
        columnVisible = true,
        Items = {},
        VisibleItems = 16, -- defaults
        Focused = false,
        CaptionLeft = "",
        CaptionRight = "",
        Label = "",
        Color = SColor.HUD_Freemode,
        Parent = nil -- BaseTab inherited tabs
    }
    return setmetatable(data, PM_Column)
end

function PM_Column:visible()
    return self.Parent ~= nil and self.Parent.Visible and self.Parent.Parent ~= nil and self.Parent.Parent:Visible()
end

function PM_Column:CurrentItem()
    return self.Items[self.index]
end

function PM_Column:Index(idx)
    if idx == nil then
        return self.index
    else
        if #self.Items == 0 then return end
        self.Items[self.index]:Selected(false)
        if idx > #self.Items then
            idx = 1
        elseif idx < 1 then
            idx = #self.Items
        end
        self.index = idx
        self.Items[self.index]:Selected(true)
        if self:visible() and self.Parent.CurrentColumnIndex == self.position then
            ScaleformUI.Scaleforms._pauseMenu._pause:CallFunction("SET_COLUMN_HIGHLIGHT", self.position, self.index - 1, false, false)
        end
    end
end

function PM_Column:ColumnVisible(bool)
    if bool == nil then
        return self.columnVisible
    else
        self.columnVisible = bool
        if self:visible() then
            ScaleformUI.Scaleforms._pauseMenu._pause:CallFunction("SHOW_COLUMN", self.position, self.columnVisible)
        end
    end
end

function PM_Column:AddItem(item)
    table.insert(self.Items, item)
end

function PM_Column:Clear()
    self:ClearColumn()
end

function PM_Column:ClearColumn()
    self.Items = {}
    self.index = 1
    if self:visible() then
        ScaleformUI.Scaleforms._pauseMenu._pause:CallFunction("SET_DATA_SLOT_EMPTY", self.position)
    end
end

function PM_Column:Populate() end

function PM_Column:SetDataSlot(index) end

function PM_Column:UpdateSlot(index) end

function PM_Column:AddSlot(index) end

function PM_Column:RemoveSlot(idx)
    if idx > #self.Items then return end
    self.Items[idx]:Selected(false)
    local curItem = self.index
    table.remove(self.Items, idx)
    if self:visible() then
        ScaleformUI.Scaleforms._pauseMenu._pause:CallFunction("REMOVE_SLOT", self.position, idx - 1)
    end
    if #self.Items > 0 then
        if idx == self.index then
            if idx > #self.Items then
                self.index = #self.Items
            elseif idx > 1 and idx <= #self.Items then
                self.index = idx
            else
                self.index = 1
            end
        else
            if curItem <= #self.Items then
                self.index = curItem
            else
                self.index = #self.Items
            end
        end
        self.Items[self.index]:Selected(true)
        if self:visible() and self.Parent.CurrentColumnIndex == self.position then
            ScaleformUI.Scaleforms._pauseMenu._pause:CallFunction("SET_COLUMN_HIGHLIGHT", self.index - 1, false, false)
        end
    end
end

function PM_Column:GoUp() end

function PM_Column:GoDown() end

function PM_Column:GoLeft() end

function PM_Column:GoRight() end

function PM_Column:Select() end

function PM_Column:GoBack() end

function PM_Column:MouseScroll(dir) end

function PM_Column:HighlightColumn(highlighted, moveFocus, prevHighlight)
    if highlighted == nil then highlighted = false end
    if moveFocus == nil then moveFocus = false end
    if prevHighlight == nil then prevHighlight = false end
    if self:visible() then
        ScaleformUI.Scaleforms._pauseMenu._pause:CallFunction("SET_COLUMN_FOCUS", self.position, highlighted, moveFocus,
            prevHighlight)
    end
end

function PM_Column:ShowColumn()
    if self:visible() then
        ScaleformUI.Scaleforms._pauseMenu._pause:CallFunction("DISPLAY_DATA_SLOT", self.position)
    end
end

function PM_Column:InitColumnScroll(visible, columns, scrollType, arrowPosition, override, xColOffset)
    if override == nil then override = false end
    if xColOffset == nil then xColOffset = 0.0 end
    if self:visible() then
        ScaleformUI.Scaleforms._pauseMenu._pause:CallFunction("INIT_COLUMN_SCROLL", self.position, visible, columns,
            scrollType, arrowPosition, override, xColOffset)
    end
end

function PM_Column:SetColumnScroll(...)
    if not self:visible() then return end
    if select(1, ...) ~= nil and select(2, ...) ~= nil and select(3, ...) ~= nil and select(4, ...) ~= nil and select(5, ...) ~= nil then
        local currentPosition, maxPosition, maxVisible, caption, forceInvisible, captionR = ...
        ScaleformUI.Scaleforms._pauseMenu._pause:CallFunction("SET_COLUMN_SCROLL", self.position, currentPosition,
            maxPosition, maxVisible, caption, forceInvisible, captionR or "")
            return
    elseif select(1, ...) ~= nil and type(select(1, ...)) == "number" and select(2, ...) ~= nil and type(select(2, ...)) == "number" and select(3, ...) ~= nil and type(select(3, ...)) == "number" then
        local currentPosition, maxPosition, maxVisible = ...
        maxVisible = maxVisible or -1
        ScaleformUI.Scaleforms._pauseMenu._pause:CallFunction("SET_COLUMN_SCROLL", self.position, currentPosition, maxPosition, maxVisible)
        return
    elseif select(1, ...) ~= nil and type(select(1, ...)) == "string" and select(2, ...) ~= nil and type(select(2, ...)) == "string" then
        local caption, rightC = select(1, ...), select(2, ...)
        BeginScaleformMovieMethod(Main.PauseMenu._pause.Handle, "SET_COLUMN_SCROLL")
        ScaleformMovieMethodAddParamInt(position)
        ScaleformMovieMethodAddParamInt(0)
        ScaleformMovieMethodAddParamInt(0)
        ScaleformMovieMethodAddParamInt(0)
        BeginTextCommandScaleformString("CELL_EMAIL_BCON")
        AddTextComponentSubstringPlayerName(caption)
        EndTextCommandScaleformString_2()
        ScaleformMovieMethodAddParamBool(false)
        ScaleformMovieMethodAddParamPlayerNameString(rightC)
        EndScaleformMovieMethod()
        return
    else
        local caption = select(1, ...)
        if type(caption) == "string" then
            BeginScaleformMovieMethod(ScaleformUI.Scaleforms._pauseMenu._pause.handle, "SET_COLUMN_SCROLL")
            ScaleformMovieMethodAddParamInt(position)
            ScaleformMovieMethodAddParamInt(0)
            ScaleformMovieMethodAddParamInt(0)
            ScaleformMovieMethodAddParamInt(0)
            BeginTextCommandScaleformString(caption)
            
            -- Process remaining arguments
            local args = {...}
            for i = 2, #args do
                local arg = args[i]
                local argType = type(arg)
                if argType == "number" then
                    if math.type(arg) == "integer" then
                        AddTextComponentInteger(arg)
                    else
                        AddTextComponentFloat(arg, 2)
                    end
                elseif argType == "string" then
                    AddTextComponentSubstringPlayerName(arg)
                end
            end
            
            EndTextCommandScaleformString_2()
            EndScaleformMovieMethod()
        end
    end
end